File: forza/cli/maintenance.py
from __future__ import annotations

import argparse
import json
import sqlite3
from pathlib import Path

from ..config import load_config
from ..logging_setup import setup_logging
from ..application import DatabaseService
from ..application.reference_seed import seed_initial_reference_text_files


def cmd_db_status(args: argparse.Namespace) -> None:
    cfg = load_config(args.config)
    setup_logging(cfg.log_file, debug=args.debug)
    # DatabaseService.status() is read-only — does not create the database.
    with DatabaseService(cfg.database_file) as database:
        status = database.status()

    print(f"Database: {status.database_file}")
    print(f"Exists:   {status.database_exists}")
    print(f"Schema:   {status.schema_state}")
    print(f"Revision: {status.current_revision or '(none)'}")
    print(f"Head:     {status.head_revision or '(none)'}")
    print("")
    print("Relational store")
    print(f"  image_files     : {status.image_files}")
    print(f"  extraction_runs : {status.extraction_runs}")
    print(f"  extraction_results: {status.extraction_results}")
    print(f"  extraction_attempts: {getattr(status, 'extraction_attempts', 0)}")
    print(f"  lap_records       : {status.lap_records}")
    print(f"  review_cases      : {status.review_cases}")
    print(f"  image_flags       : {status.image_flags}")
    print(f"  export_artifacts  : {status.export_artifacts}")
    print(f"  reference_tracks  : {getattr(status, 'reference_tracks', 0)}")
    print(f"  reference_cars    : {getattr(status, 'reference_cars', 0)}")
    print(f"  external_imports  : {getattr(status, 'external_record_imports', 0)}")
    print(f"  external_laps     : {getattr(status, 'external_lap_records', 0)}")


def cmd_db_reset(args: argparse.Namespace) -> None:
    """Delete the configured SQLite database and sidecar files."""
    if not args.yes:
        raise SystemExit("Refusing to reset database without --yes")

    cfg = load_config(args.config)
    database_file = Path(cfg.database_file)

    if database_file.exists():
        _ensure_exclusive_access(database_file)

    wal_file = Path(f"{database_file}-wal")
    shm_file = Path(f"{database_file}-shm")
    if wal_file.exists() or shm_file.exists():
        print(
            "WARNING: WAL/SHM sidecar file(s) present — this can mean a "
            "connection had the database open recently. Proceeding since no "
            "connection currently holds an exclusive lock."
        )

    targets = [database_file, wal_file, shm_file]
    removed: list[Path] = []
    for target in targets:
        if target.exists():
            target.unlink()
            removed.append(target)

    print("Database reset")
    print(f"Removed: {len(removed)} file(s)")
    for target in removed:
        print(f"  - {target}")


def _ensure_exclusive_access(database_file: Path) -> None:
    """Abort instead of deleting if another connection currently holds the database.

    Deleting a SQLite file out from under an open connection can corrupt
    whatever that connection was mid-write on (see audit finding C-2).
    Requesting an EXCLUSIVE lock surfaces an in-use database as a clear error
    instead of a silent race. A file that fails to open because it isn't a
    valid SQLite database at all (``sqlite3.DatabaseError``) is a legitimate
    reason to run db-reset in the first place, so that case is not blocked —
    only an actual lock held by another connection is (``sqlite3.OperationalError``).
    """
    conn = sqlite3.connect(str(database_file), timeout=0)
    try:
        conn.execute("PRAGMA locking_mode=EXCLUSIVE")
        conn.execute("BEGIN EXCLUSIVE")
        conn.execute("COMMIT")
    except sqlite3.OperationalError as exc:
        raise SystemExit(
            f"Refusing to reset database: {database_file} appears to be in use by "
            f"another connection ({exc}). Close any running Forza processes (GUI, "
            "CLI runs, or scripts) and try again."
        ) from exc
    except sqlite3.DatabaseError:
        # Not a valid SQLite file at all — resetting is the legitimate fix.
        pass
    finally:
        conn.close()


def cmd_db_upgrade(args: argparse.Namespace) -> None:
    """Apply all pending Alembic migrations."""
    from ..db.migrate import DatabaseSchemaState, detect_database_state, upgrade_database

    cfg = load_config(args.config)
    setup_logging(cfg.log_file, debug=getattr(args, "debug", False))

    state = detect_database_state(cfg.database_file)
    if state == DatabaseSchemaState.UNMANAGED:
        print(f"ERROR: Unmanaged database detected at {cfg.database_file}")
        print("       It has tables but no alembic_version; Alembic cannot manage it.")
        print("       Run: python -m forza maintenance db-reset --yes")
        print("       Then retry: python -m forza maintenance db-upgrade")
        raise SystemExit(1)

    print(f"Upgrading database: {cfg.database_file}")
    upgrade_database(cfg.database_file)
    added_tracks, added_cars = seed_initial_reference_text_files(cfg.database_file)
    print(f"Seeded references: {added_tracks} track(s), {added_cars} car(s) added.")
    print("Done.")


def cmd_db_doctor(args: argparse.Namespace) -> None:
    """Report relational integrity issues that matter before release/reruns."""
    from ..application import DbDoctorService

    cfg = load_config(args.config)
    setup_logging(cfg.log_file, debug=getattr(args, "debug", False))
    report = DbDoctorService().run(cfg.database_file)
    if getattr(args, "json", False):
        print(json.dumps({
            "database_file": str(report.database_file),
            "schema_state": report.schema_state,
            "ok": report.ok,
            "checks": [
                {
                    "key": check.key,
                    "severity": check.severity,
                    "count": check.count,
                    "detail": check.detail,
                    "ok": check.ok,
                }
                for check in report.checks
            ],
        }, indent=2))
    else:
        _print_db_doctor_report(report)
    if not report.ok:
        raise SystemExit(2)


def _print_db_doctor_report(report) -> None:
    print(f"Database: {report.database_file}")
    print(f"Schema:   {report.schema_state}")
    print(f"OK:       {report.ok}")
    for check in report.checks:
        status = "OK" if check.ok else check.severity.upper()
        print(f"[{status}] {check.key}: {check.count} - {check.detail}")


def cmd_config_check(args: argparse.Namespace) -> None:
    """Validate forza_config.ini and report any errors."""
    from ..config import validate_config
    from ..exceptions import ConfigValidationError

    try:
        cfg = load_config(args.config)
    except Exception as exc:
        print(f"ERROR: Could not load config: {exc}")
        raise SystemExit(1)

    try:
        validate_config(cfg)
        print(f"Configuration is valid. ({args.config})")
    except ConfigValidationError as exc:
        print(str(exc))
        raise SystemExit(1)

File: forza-rust/crates/forza-cli/src/commands/maintenance.rs
//! Database inspection commands (`maintenance ...`).

use std::path::{Path, PathBuf};

use rusqlite::Connection;

use super::common::table_count;
use crate::APP_VERSION;
use forza_db::doctor;
use forza_db::migration::{SchemaStatus, schema_status};

pub(crate) fn cmd_db_status(db_path: &Path) -> anyhow::Result<()> {
    let status = schema_status(db_path)?;
    let label = match status {
        SchemaStatus::Empty => "empty",
        SchemaStatus::Current => "current",
        SchemaStatus::Incompatible { found } => {
            println!("database_file = {}", db_path.display());
            println!("schema_state  = incompatible (user_version={found})");
            std::process::exit(1);
        }
    };

    println!("database_file = {}", db_path.display());
    println!("schema_state  = {label}");

    if status == SchemaStatus::Empty {
        std::process::exit(1);
    }

    let conn = forza_db::open_connection(db_path)?;

    let image_files = table_count(&conn, "image_files");
    let extraction_runs = table_count(&conn, "extraction_runs");
    let extraction_results = table_count(&conn, "extraction_results");
    let extraction_attempts = table_count(&conn, "extraction_attempts");
    let lap_records = table_count(&conn, "lap_records");
    let review_cases = table_count(&conn, "review_cases");
    let review_corrections = table_count(&conn, "review_corrections");
    let image_flags = table_count(&conn, "image_flags");
    let export_artifacts = table_count(&conn, "export_artifacts");
    let reference_tracks = table_count(&conn, "reference_tracks");
    let reference_cars = table_count(&conn, "reference_cars");
    let external_record_imports = table_count(&conn, "external_record_imports");
    let external_lap_records = table_count(&conn, "external_lap_records");

    fn show(count: Option<i64>) -> String {
        count
            .map(|n| n.to_string())
            .unwrap_or_else(|| "ERR (unreadable)".to_string())
    }
    println!();
    println!("Relational store");
    println!("  image_files         : {}", show(image_files));
    println!("  extraction_runs     : {}", show(extraction_runs));
    println!("  extraction_results  : {}", show(extraction_results));
    println!("  extraction_attempts : {}", show(extraction_attempts));
    println!("  lap_records         : {}", show(lap_records));
    println!("  review_cases        : {}", show(review_cases));
    println!("  review_corrections  : {}", show(review_corrections));
    println!("  image_flags         : {}", show(image_flags));
    println!("  export_artifacts    : {}", show(export_artifacts));
    println!("  reference_tracks    : {}", show(reference_tracks));
    println!("  reference_cars      : {}", show(reference_cars));
    println!(
        "  external_record_imports : {}",
        show(external_record_imports)
    );
    println!("  external_lap_records    : {}", show(external_lap_records));

    Ok(())
}

pub(crate) fn cmd_db_doctor(db_path: &Path, json: bool) -> anyhow::Result<()> {
    let schema_label = report_schema_status(db_path)?;
    let report = if schema_label == "empty" {
        // Empty/missing DB — use the lightweight doctor that doesn't query tables.
        doctor::doctor_on_path(db_path)?
    } else {
        // Schema present — run the full battery.
        doctor::run_full_doctor(&forza_db::open_connection(db_path)?, schema_label)?
    };
    if json {
        let checks: Vec<_> = report
            .checks
            .iter()
            .map(|check| {
                serde_json::json!({
                    "key": check.key,
                    "severity": match check.severity {
                        doctor::DoctorSeverity::Error => "error",
                        doctor::DoctorSeverity::Warning => "warning",
                        doctor::DoctorSeverity::Info => "info",
                    },
                    "count": check.count,
                    "detail": check.detail,
                    "ok": check.ok,
                })
            })
            .collect();
        println!(
            "{}",
            serde_json::to_string_pretty(&serde_json::json!({
                "database_file": db_path,
                "schema_state": report.schema_status,
                "ok": report.ok,
                "checks": checks,
            }))?
        );
    } else {
        println!("forza {APP_VERSION}");
        println!("Database: {}", db_path.display());
        println!("Schema:   {}", report.schema_status);
        println!("OK:       {}", report.ok);
        for check in &report.checks {
            let status = if check.ok {
                "OK"
            } else {
                match check.severity {
                    doctor::DoctorSeverity::Error => "ERROR",
                    doctor::DoctorSeverity::Warning => "WARN",
                    doctor::DoctorSeverity::Info => "INFO",
                }
            };
            let count = check.count;
            println!(
                "[{status}] {key}: {count} - {detail}",
                key = check.key,
                detail = check.detail
            );
        }
    }
    if !report.ok {
        std::process::exit(2);
    }
    Ok(())
}

fn report_schema_status(db_path: &Path) -> anyhow::Result<String> {
    let status = schema_status(db_path)?;
    Ok(match status {
        SchemaStatus::Empty => "empty".to_string(),
        SchemaStatus::Current => "current".to_string(),
        SchemaStatus::Incompatible { found } => format!("incompatible (user_version={found})"),
    })
}

/// Verify no other connection holds the database by requesting an EXCLUSIVE lock.
fn ensure_exclusive_access(db_path: &Path) -> anyhow::Result<()> {
    let conn = Connection::open(db_path)?;
    // Try BEGIN EXCLUSIVE; COMMIT to acquire the lock.
    // On Windows the bundled SQLite may reject locking_mode=EXCLUSIVE pragma,
    // so we fall back to just BEGIN EXCLUSIVE and catch SQLITE_BUSY.
    match conn.execute("BEGIN EXCLUSIVE", []) {
        Ok(_) => {
            // COMMIT on the SAME connection that began the transaction: a
            // COMMIT on a second connection is a no-op and would hold the
            // lock until `conn` drops (longer than intended).
            let _ = conn.execute("COMMIT", []);
            Ok(())
        }
        Err(e) => {
            let msg = format!("{e}");
            if msg.contains("not a database") || msg.contains("file is not a database") {
                // Not a valid SQLite file — resetting is the legitimate fix.
                Ok(())
            } else if msg.contains("database is locked")
                || msg.contains("locked")
                || msg.contains("BUSY")
            {
                Err(anyhow::anyhow!(
                    "Refusing to reset database: {} appears to be in use by another connection (database locked). Close any running Forza processes (GUI, CLI runs, or scripts) and try again.",
                    db_path.display()
                ))
            } else {
                Err(anyhow::anyhow!(
                    "Refusing to reset database: {} appears to be in use by another connection ({e}). Close any running Forza processes (GUI, CLI runs, or scripts) and try again.",
                    db_path.display()
                ))
            }
        }
    }
}

pub(crate) fn cmd_db_reset(db_path: &Path, yes: bool) -> anyhow::Result<()> {
    let sidecars = [
        PathBuf::from(format!("{}-wal", db_path.display())),
        PathBuf::from(format!("{}-shm", db_path.display())),
    ];

    if db_path.exists() {
        ensure_exclusive_access(db_path)?;
    }

    // Check for stale WAL/SHM sidecars after exclusive lock check
    for sidecar in &sidecars {
        if sidecar.exists() {
            let name = sidecar
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown");
            println!(
                "WARNING: {name} sidecar file present — a connection may have held the database recently.",
            );
        }
    }

    println!("This deletes:");
    for path in std::iter::once(db_path.to_path_buf()).chain(sidecars.iter().cloned()) {
        if path.exists() {
            println!("  {}", path.display());
        }
    }
    if !yes {
        // Python hard-errors here (SystemExit): a bare db-reset must fail
        // loudly, not exit 0 as if the reset happened.
        return Err(anyhow::anyhow!(
            "refusing to delete without --yes (re-run with --yes to confirm)"
        ));
    }
    for path in std::iter::once(db_path.to_path_buf()).chain(sidecars.iter().cloned()) {
        if path.exists() {
            std::fs::remove_file(&path)?;
            println!("removed {}", path.display());
        }
    }
    Ok(())
}
